design_drafts/joint_interfaces_and_transmissions.md

```diff
@@ -0,0 +1,145 @@
+This however already shows a shortcomings of this design as a whole as with a fixed set of interfaces, a fixed set of functions were created, most of them defined empty, for example:
+```
+  virtual void jointToActuatorEffort(const JointData&    jnt_data,
+                                           ActuatorData& act_data) = 0;
+```
+All transmissions inheriting this interface has to define empty functions for those it does not operate on. Additionally, this set of functions is not extendable in a clean way, without incurring additional interface inheritance (and a general mess...) on the derived classes.
+ 
+## Scope:
+This design document proposes changes to 
+1) joint and actuator interfaces
+2) transmission parsing
+3) controller interfaces (as a consequence of 1)
+
+and ties in with the use of the [flexible joint states message](https://github.com/ros-controls/roadmap/blob/master/design_drafts/flexible_joint_states_msg.m) 
+4) in the context of replacing `ActuatorData` and `JointData` from `ros_control` on ROS Melodic in how transmission loading and runtime is handled
+5) the `joint_state_controller` which depends on the "read-only" `JointStateHandle` class to report position, velocity and effort values.
+
+## Proposal
+
+### urdf
+Over the years we didn't receive any feedback about the URDF notation being good/bad which means it's not currently a pain point for anyone at the moment, this design will not propose any changes to that.
+Let's stick with the notation:
+```
+<transmission name="extended_simple_trans">
+    <type>transmission_interface/ExtendedSimpleTransmission</type>
+    <joint name="joint1">
+      <offset>0.0</offset>
+      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
+      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
+      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
+      <hardwareInterface>awesome_interface/FooJointInterface</hardwareInterface>
+      <role></role>
+    </joint>
+    <actuator name="actuator1">
+      <mechanicalReduction>50</mechanicalReduction>
+    </actuator>
+</transmission>
+```
+The content of a `hardwareInterface`  XML attribute should match up with strings provided by `ros2_control` in the shape of familiar names:
+`hardware_interface::PositionJointInterface, hardware_interface::VelocityJointInterface, hardware_interface::EffortJointInterface`.
+By relying on strings internally in the framework but using them mostly through shared names, a standard nomenclature is kept all the while allowing custom extensions in an easy and transparent way.
+Note the in the example above, since all names are merely strings, `awesome_interface/FooJointInterface` - a custom interface - can be used if it is provided by the `RobotHardware`.
+
+### RobotHardware
+
+```
+void RobotHardware::declareInterface(string name);
```

Comment:
We should check how to do this exactly. This is a possible way, but for registering new `Actuators` which than have an instance of `Transmission` class in them.